presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java

```diff
@@ -343,26 +342,7 @@ public HiveWriter createWriter(Page partitionColumns, int position, OptionalInt
                 throw new PrestoException(HIVE_PARTITION_READ_ONLY, "Cannot insert into an existing partition of Hive table: " + partitionName.get());
             }
             updateMode = APPEND;
-
-            // Check the column types in partition schema match the column types in table schema
-            List<Column> tableColumns = table.getDataColumns();
-            List<Column> existingPartitionColumns = partition.get().getColumns();
-            for (int i = 0; i < min(existingPartitionColumns.size(), tableColumns.size()); i++) {
-                HiveType tableType = tableColumns.get(i).getType();
-                HiveType partitionType = existingPartitionColumns.get(i).getType();
-                if (!tableType.equals(partitionType)) {
-                    throw new PrestoException(HIVE_PARTITION_SCHEMA_MISMATCH, format("" +
-                                    "There is a mismatch between the table and partition schemas. " +
-                                    "The column '%s' in table '%s' is declared as type '%s', " +
-                                    "but partition '%s' declared column '%s' as type '%s'.",
-                            tableColumns.get(i).getName(),
-                            tableName,
-                            tableType,
-                            partitionName,
-                            existingPartitionColumns.get(i).getName(),
-                            partitionType));
-                }
-            }
+            checkTableAndPartitionSchemaMatch(table, tableName, partition.get(), partitionName.get());
```

Comment:
1. I was confused by this at the beginning. How come this is checking type equality when Presto has supported type evolution. Then I realize this code is attempting to insert into an existing partition. This is a very different problem. It would be nice if the error message a few lines below be more explicit on this. Something along the lines of `You are trying to write into an existing partition in a table. The table schema has changed since the creation of the partition. Insertion into such partition is not supported. The column ...`
2. In OVERWRITE mode, there is no underlying technical reason to justify this check. We should not make this check.

As a result, it seems unnecessary to turn this into a method. In addition, the particularity of the error message might make it undesirable to do so.